lib/Canonicalize/SimplifySwitch/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,234 @@
+                                                     DT,
+                                                     Limits,
+                                                     Oracle::AdvancedValueInfo);
+      auto &DFG = Results.dataFlowGraph();
+      // Identify the only node in the data flow graph that's associated to an
+      // Oracle and use it as a switch condition.
+      bool BailOut = false;
+      BidirectionalNode<DataFlowNode> *StartNode = nullptr;
+
+      for (auto BidirectNode : DFG.nodes()) {
+        const DataFlowNode &Node = *BidirectNode;
+        // Ignore nodes that are constants in the DFG.
+        if (llvm::isa<llvm::ConstantInt>(Node.Value))
+          continue;
+
+        if (Node.UseOracle) {
+          if (StartNode != nullptr) {
+            BailOut = true;
+            break;
+          } else {
+            // TODO: Fix this const_cast.
+            auto BiNode = const_cast<BiDFNode *>(BidirectNode);
+            StartNode = BiNode;
+          }
+        }
+      }
+
+      if (StartNode == nullptr or BailOut)
+        continue;
+
+      if (not Results.values())
+        continue;
+
+      auto StartNodeInst = dyn_cast<llvm::Instruction>(StartNode->Value);
+      if (StartNodeInst and isa<llvm::PHINode>(StartNodeInst))
+        continue;
+      auto ConditionInst = dyn_cast<llvm::Instruction>(Condition);
+      if (ConditionInst and isa<llvm::PHINode>(ConditionInst))
+        continue;
+
+      if (StartNode->Value == Condition)
+        continue;
+
+      if (Results.values()->size() != Switch->getNumCases()) {
+        revng_log(SwitchOptLogger, "Unexpected values!");
+        continue;
+      }
+
+      std::map<llvm::ConstantInt *, llvm::BasicBlock *> NewLabels;
+      for (auto &Value : *StartNode->OracleRange) {
+        auto Old = DFG.materializeOne(StartNode, MO, Value);
+        if (not Old)
+          return false;
+        auto *ConstantForOld = toLLVMConstant(Switch->getContext(),
+                                              (*Old).value());
+        auto *ConstantForTheValue = toLLVMConstant(Switch->getContext(), Value);
+
+        auto *BlockForValue = getBlockFor(Switch, ConstantForOld);
+
+        if (not BlockForValue)
+          return false;
+        NewLabels[cast<llvm::ConstantInt>(ConstantForTheValue)] = BlockForValue;
+      }
+
+      // Recreate the new switch.
+      llvm::IRBuilder<> Builder(Switch);
+      auto *UnreachBB = llvm::BasicBlock::Create(Switch->getContext(),
+                                                 "unreachable",
+                                                 &F);
+      llvm::IRBuilder<> IRBUnreachable(UnreachBB);
+      IRBUnreachable.CreateUnreachable();
+      auto *NewSwitch = Builder.CreateSwitch(StartNode->Value,
+                                             UnreachBB,
+                                             NewLabels.size());
+
+      for (auto &NewLabel : NewLabels) {
+        NewSwitch->addCase(NewLabel.first, NewLabel.second);
+      }
```

Comment:
1. Confusing code organization.
2. Do not create an `IRBuilder` to emit a single instruction.